### ISSUE TITLE
conda env update --prune does not remove installed packages not defined in environment.yml

### ISSUE THREAD

**Issue by golubitsky** [conda/conda#7279]
**I'm submitting a...**
  - [x] bug report
  - [ ] feature request

## Current Behavior
`conda env update --prune` does not remove packages no longer referenced in the `environment.yml` file.

### Steps to Reproduce

Define an `environment.yml` file in a project directory as follows:

    name: new-project-env
    dependencies:
    - python=3
    - numpy

In the project directory run the following commands:

    conda env create
    conda activate new-project-env
    python
    import numpy
    exit()

This succeeds as expected.
Update the `environment.yml` file as follows, **removing numpy**,

    name: new-project-env
    dependencies:
    - python=3

Run the following commands in the project directory:

    conda deactivate
    conda env update --prune
    conda activate new-project-env
    python
    import numpy
    exit()

**This numpy import succeeds and seems like a bug**. 

## Expected Behavior
Per [the docs](https://conda.io/docs/commands/env/conda-env-update.html) (--prune means "remove installed packages not defined in environment.yml") I expect that after installing a package (numpy in the example above) and subsquently removing it from `environment.yml`  the package should be removed from the relevant conda environment.

## Current Workaround
I've found that we can use `conda env create --force` to overwrite the existing environment and thus ensure that the environment accurately reflects the `environment.yml` file. But this requires recreating the environment from scratch.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : new-project-env
    active env location : /home/mikhailgolubitsky/anaconda3/envs/new-project-env
            shell level : 2
       user config file : /home/mikhailgolubitsky/.condarc
 populated config files : /home/mikhailgolubitsky/.condarc
          conda version : 4.5.3
    conda-build version : 3.0.27
         python version : 3.6.3.final.0
       base environment : /home/mikhailgolubitsky/anaconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /home/mikhailgolubitsky/anaconda3/pkgs
                          /home/mikhailgolubitsky/.conda/pkgs
       envs directories : /home/mikhailgolubitsky/anaconda3/envs
                          /home/mikhailgolubitsky/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.3 requests/2.18.4 CPython/3.6.3 Linux/4.13.0-39-generic ubuntu/17.10 glibc/2.26
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
==> /home/mikhailgolubitsky/.condarc <==
channels:
  - conda-forge
  - defaults
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
# packages in environment at /home/mikhailgolubitsky/anaconda3/envs/new-project-env:
#
# Name                    Version                   Build  Channel
ca-certificates           2018.4.16                     0    conda-forge
certifi                   2018.4.16                py36_0    conda-forge
intel-openmp              2018.0.0                      8    defaults
libgcc-ng                 7.2.0                hdf63c60_3    defaults
libgfortran-ng            7.2.0                hdf63c60_3    defaults
mkl                       2018.0.2                      1    defaults
mkl_fft                   1.0.2                    py36_0    conda-forge
mkl_random                1.0.1                    py36_0    conda-forge
ncurses                   5.9                          10    conda-forge
numpy                     1.14.3           py36h14a74c5_0    defaults
numpy-base                1.14.3           py36hdbf6ddf_0    defaults
openssl                   1.0.2o                        0    conda-forge
pip                       9.0.3                    py36_0    conda-forge
python                    3.6.5                         1    conda-forge
readline                  7.0                           0    conda-forge
setuptools                39.1.0                   py36_0    conda-forge
sqlite                    3.20.1                        2    conda-forge
tk                        8.6.7                         0    conda-forge
wheel                     0.31.0                   py36_0    conda-forge
xz                        5.2.3                         0    conda-forge
zlib                      1.2.11                        0    conda-forge


```
</p></details>

**Comment by ShannonGreen** [comment]
This is the same root cause as #6809. The problem is that conda looks through its history file and sees that you've requested those specs in the past, so it adds them back.

In my opinion, the fix would either be for `conda env update` to not take specs from the history at all, or for `conda env` operations to not populate the history at all, or both.

If you'd like a workaround that is faster than `conda env create --force` you can truncate the history file located at `<env>/conda-meta/history` (the file needs to be present but can be empty).

**Comment by einarpersson** [comment]
This confused me a lot also.

**Comment by prusse-martin** [comment]
The ```truncate the history file located at `<env>/conda-meta/history` ``` hack does not work any longer for me.
I was testing conda 4.7.5 (windows and linux) when stumbled upon this.

**Comment by nicoddemus** [comment]
> In my opinion, the fix would either be for conda env update to not take specs from the history at all, or for conda env operations to not populate the history at all, or both.

Seems like the correct approach, IMHO. Any comments from the team? We are willing to work on this and provide a PR, as we dearly depend on this and is preventing us from upgrading.

**Comment by msarahan** [comment]
I'm not sure I fully understand the desired behavior, but yes, absolutely, a PR to implement better behavior is welcome.

The party I don't understand is why basing it on the history is bad. It seems like the history of explicit specs would be there best source of minimal information for env specs, but I must be missing the reasons why that doesn't work in practice. Perhaps the reason is that the removal part of the update is not factored in correctly?

**Comment by nicoddemus** [comment]
> I'm not sure I fully understand the desired behavior, but yes, absolutely, a PR to implement better behavior is welcome.

Sure, let me explain with an example.

## conda <4.4

Suppose we have this `environment.yaml` file:

```yaml
name: env-app

dependencies:
- python
- pytest-mock
```

When running `conda env update` for the 1st time, conda creates an environment with `python`, `pytest-mock`, and their dependencies, including `pytest`, which `pytest-mock` depends on. 

If we change the `environment.yaml` to:

```yaml
name: env-app

dependencies:
- python
```

and execute `conda env update` again, nothing happens, because by default it will only add to the environment, never remove.

When we implemented `--prune` originally, its purpose was to remove every package and dependency not listed in the `environment.yaml` file. 

So adding `--prune` to the 2nd run, conda would uninstall `pytest-mock` and `pytest`, as they are no longer listed.

## conda 4.4+

This has changed since version 4.4 with the introduction of the history feature: conda keeps track of what was installed previously, so users typing `conda install` multiple times would get a consistent environment (that's my understanding of the feature).

But this has also affected `conda env update --prune`, most likely unintentionally (I suspect the reason is that it uses `conda install` behind the scenes).

So what happens in the 2nd run now, using `--prune`, is that `pytest-mock` and `pytest` are kept in the environment.

So effectively `--prune` has been broken since 4.4, as it no longer removes previous packages.


> The party I don't understand is why basing it on the history is bad.

The main use case is to get the exact environment described in the yaml file, exactly what you would get with `conda create --force`, but much quicker as you don't have to delete/re-link everything.

**Comment by nicoddemus** [comment]
@msarahan are you OK with a PR fixing the behavior of `--prune` as I describe above?

**Comment by msarahan** [comment]
I see, thanks for the explanation.  Yes, any PR that fixes this is welcome. It's not immediately obvious to me how to do it. Perhaps a flag like --replace-specs that ignores the history and only keeps the specs that are passed explicitly.

The implementation of history specs being added in is in conda/core/solve.py in the _add_specs method. Let me know if you have any questions.

**Comment by nicoddemus** [comment]
Thanks for the pointers!

**Comment by selik** [comment]
Can we raise a warning or `NotImplementedError` for the update `--prune` option until it is fixed?

**Comment by kahing** [comment]
I am having trouble with the workaround as well (truncating the history file). Step to reproduce (conda 4.8.2):

1. conda create --clone base -n update-test
2. conda activate update-test
3. conda env export > orig.yml
4. conda install ipython
5. truncate --size 0 ~/miniconda3/envs/update-test/conda-meta/history
6. conda env update --prune -f orig.yml

after those steps ipython is still installed.

**Comment by dandiez** [comment]
Does anybody know any currently working workaround for this bug that does not involve recreating the environment from scratch (which is too slow)?

**Comment by prusse-martin** [comment]
We are using a patched conda (version 4.8.3) at the moment for this:
https://github.com/conda/conda/pull/9614

**Comment by chrisyeh96** [comment]
Has there been any recent progress on this issue?

**Comment by prusse-martin** [comment]
No.
I did not have time to check a recent version of conda and if the patch proposed in #9614 even applies to recent versions (given the age of this I will expect it to not).

**Comment by j-neefs** [comment]
fyi workaround of truncating the history file also not working for me on version 4.10.1. Preferring for now to keep the history file and optimize the conda env update speed until this is fixed

**Comment by rcasero** [comment]
This is supposed to be about to be fixed in #9614

**Comment by mfansler** [comment]
One workaround is that `mamba env update --prune` works fine. Simple demonstration in this thread: https://stackoverflow.com/a/75176934/570918
